File: ios/Classes/LibreLocationLogger.swift
import Foundation
import os.log

/// Structured logging system for iOS native layer.
/// Uses os_log with subsystem "io.rezivure.libre_location" and stores
/// the last N entries in memory for retrieval via getLog().
final class LibreLocationNativeLogger {

    static var logLevel: Int = 0 // 0=off, 1=error, 2=warning, 3=info, 4=debug, 5=verbose
    private static let maxEntries = 500
    private static var entries: [[String: Any]] = []
    private static let lock = NSLock()

    private static let subsystem = "io.rezivure.libre_location"
    private static let category = "location"

    @available(iOS 10.0, *)
    private static let osLog = OSLog(subsystem: subsystem, category: category)

    private static let dateFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    static func verbose(_ message: String) { log(5, "verbose", message) }
    static func debug(_ message: String) { log(4, "debug", message) }
    static func info(_ message: String) { log(3, "info", message) }
    static func warning(_ message: String) { log(2, "warning", message) }
    static func error(_ message: String) { log(1, "error", message) }

    private static func log(_ level: Int, _ levelName: String, _ message: String) {
        guard logLevel > 0, level <= logLevel else { return }

        // Write to os_log
        if #available(iOS 10.0, *) {
            let type: OSLogType
            switch level {
            case 1: type = .error
            case 2: type = .default  // warning
            case 3: type = .info
            case 4: type = .debug
            default: type = .debug
            }
            os_log("%{public}@", log: osLog, type: type, message)
        }

        // Store in ring buffer
        let entry: [String: Any] = [
            "timestamp": dateFormatter.string(from: Date()),
            "level": levelName,
            "message": message,
            "platform": "ios",
        ]

        lock.lock()
        entries.append(entry)
        if entries.count > maxEntries {
            entries.removeFirst()
        }
        lock.unlock()
    }

    static func getLog() -> [[String: Any]] {
        lock.lock()
        let result = entries
        lock.unlock()
        return result
    }

    static func clear() {
        lock.lock()
        entries.removeAll()
        lock.unlock()
    }
}

File: android/src/main/kotlin/io/rezivure/libre_location/LibreLocationLogger.kt
package io.rezivure.libre_location

import android.util.Log
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.ConcurrentLinkedDeque

/**
 * Structured logging system for the Android native layer.
 * Uses android.util.Log with tag "LibreLocation" and respects the configured log level.
 * Stores the last N entries in memory for retrieval via getLog().
 */
object LibreLocationLogger {
    private const val TAG = "LibreLocation"
    private const val MAX_ENTRIES = 500

    // LogLevel: 0=off, 1=error, 2=warning, 3=info, 4=debug, 5=verbose
    @Volatile
    var logLevel: Int = 0

    private val entries = ConcurrentLinkedDeque<Map<String, Any>>()
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US)

    fun verbose(message: String) = log(5, "verbose", message)
    fun debug(message: String) = log(4, "debug", message)
    fun info(message: String) = log(3, "info", message)
    fun warning(message: String) = log(2, "warning", message)
    fun error(message: String) = log(1, "error", message)

    private fun log(level: Int, levelName: String, message: String) {
        if (logLevel == 0 || level > logLevel) return

        // Write to android.util.Log
        when (level) {
            1 -> Log.e(TAG, message)
            2 -> Log.w(TAG, message)
            3 -> Log.i(TAG, message)
            4 -> Log.d(TAG, message)
            5 -> Log.v(TAG, message)
        }

        // Store in ring buffer
        val entry = mapOf<String, Any>(
            "timestamp" to dateFormat.format(Date()),
            "level" to levelName,
            "message" to message,
            "platform" to "android",
        )
        entries.addLast(entry)
        while (entries.size > MAX_ENTRIES) {
            entries.pollFirst()
        }
    }

    fun getLog(): List<Map<String, Any>> = entries.toList()

    fun clear() = entries.clear()
}
